cdap-spark-core-base/src/main/java/io/cdap/cdap/app/runtime/spark/submit/AbstractSparkSubmitter.java

```diff
@@ -105,11 +105,15 @@ protected void cancelTask() {
     executor.submit(new Runnable() {
       @Override
       public void run() {
-        List<String> extraArgs = beforeSubmit();
         try {
+          List<String> extraArgs = beforeSubmit();
           String[] submitArgs = Iterables.toArray(Iterables.concat(args, extraArgs), String.class);
           submit(runtimeContext, submitArgs);
```

Comment:
I'm trying to understand how this works in the Hadoop case but not in k8s. 

In Hadoop, SparkSubmit.main() throws an exception if it fails but on k8s it doesn't?

Reply:
Correct, the reason is current spark implementation on k8s where driver pod status in client mode doesnt return exception upon failed driver pod status. 

https://github.com/apache/spark/blob/master/resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/submit/LoggingPodStatusWatcher.scala#L64